### PR TITLE
[5.1] BelongsToThrough relation for Eloquent Model

### LINKED ISSUE #6161: BelongsToThrough relationship needed
Originally brought up and dismissed in #3052. (see for more examples)

The HasManyThrough relationship can not be used in both directions.

The simplest analogy is
person a knows person b's phone number
person b know's person c's phone number.

Person c can get person a's phone number via HasManyThrough.

Hasmanythrough does NOT work for getting from person a to person c as assumed (the reason #3052 was closed.

Related is #3786 where @taylorotwell commented  "How can hasManyThrough NOT work both ways. If it works one way, it by definition must work the "other way".
HasManyThrough cannot work this way because C does not point to B (foreign key), and B does not point to A (fk).

Currently two relationships are needed to represent this:
    A belongs to B
    B belongs to C.

### PR DESCRIPTION
**Pull request by znck**
This fixes #6161

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Model.php [hunk 1]
```diff
@@ -6,6 +6,7 @@
 use Exception;
 use ArrayAccess;
 use Carbon\Carbon;
+use Illuminate\Database\Eloquent\Relations\BelongsToThrough;
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: please sort imports by length

File: src/Illuminate/Database/Eloquent/Model.php [hunk 2]
```diff
@@ -838,6 +839,38 @@ public function belongsTo($related, $foreignKey = null, $otherKey = null, $relat
...
+     * @param string|array $through
+     * @param string|null $localKey
+     *
+     * @return \Illuminate\Database\Eloquent\Relations\Relation
+     */
+    public function belongsToThrough($related, $through, $localKey = null)
+    {
+        $related = new $related;
+        if (!($related instanceof Model)) {
+            throw new \InvalidArgumentException("\$related class should be instance of \\Iluminate\\Database\\Eloquent\\Model.");
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: please import this

File: src/Illuminate/Database/Eloquent/Model.php [hunk 3]
```diff
@@ -838,6 +839,38 @@ public function belongsTo($related, $foreignKey = null, $otherKey = null, $relat
     }
 
     /**
+     * Define an inverse one-to-many or many relationship.
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: please fix the cs in this entire phpdoc

File: src/Illuminate/Database/Eloquent/Model.php [hunk 4]
```diff
@@ -838,6 +839,38 @@ public function belongsTo($related, $foreignKey = null, $otherKey = null, $relat
...
+    public function belongsToThrough($related, $through, $localKey = null)
+    {
+        $related = new $related;
+        if (!($related instanceof Model)) {
+            throw new \InvalidArgumentException("\$related class should be instance of \\Iluminate\\Database\\Eloquent\\Model.");
+        }
+        $models = [];
+        foreach ((array)$through as $key => $model) {
+            $object = new $model;
+            if (!($object instanceof Model)) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: don't need extra brackets

File: src/Illuminate/Database/Eloquent/Model.php [hunk 5]
```diff
@@ -838,6 +839,38 @@ public function belongsTo($related, $foreignKey = null, $otherKey = null, $relat
...
+                throw new \InvalidArgumentException("\$through classes should be instance of \\Iluminate\\Database\\Eloquent\\Model.");
+            }
+            $models[] = $object;
+        }
+
+        if (empty($through)) {
+            throw new \InvalidArgumentException("\$through should contain one or more classes.");
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: i don't find it helpful to name variables in exeception messages

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+     */
+    protected $models;
+
+    /**
+     * @type string|null
+     */
+    protected $localKey;
+
+    /**
+     * BelongsToThrough constructor.
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: not useful

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 7]
```diff
@@ -0,0 +1,174 @@
+<?php namespace Illuminate\Database\Eloquent\Relations;
+
+use Illuminate\Database\Eloquent\Builder;
+use Illuminate\Database\Eloquent\Collection;
+use Illuminate\Database\Eloquent\Model;
+
+class BelongsToThrough extends Relation
+{
+    const RELATED_THROUGH_KEY = '__deep_related_through_key';
+    /**
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: missing space

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+     * Get the results of the relationship.
+     *
+     * @return mixed
+     */
+    public function getResults()
+    {
+        return $this->getQuery()->first();
+    }
+
+    /**
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: missing short description

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+            }
+        }
+
+        return $models;
+    }
+
+    /**
+     * Get the results of the relationship.
+     *
+     * @return mixed
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: more specific?

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+        foreach ($this->models as $key => $model) {
+            $other = $model->getTable() . '.' . $prev;
+            $this->getQuery()->leftJoin($model->getTable(), $one, '=', $other);
+
+            $prev = $model->getForeignKey();
+            $one = $model->getQualifiedKeyName();
+        }
+    }
+
+    /**
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: again

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+     * @return bool
+     */
+    public function parentSoftDeletes()
+    {
+        return in_array('Illuminate\Database\Eloquent\SoftDeletes',
+            class_uses_recursive(get_class($this->getParent())));
+    }
+
+    /**
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: again

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+        $dictionary = [];
+
+        foreach ($results as $result) {
+            $dictionary[$result->{static::RELATED_THROUGH_KEY}] = $result;
+            $result->offsetUnset(static::RELATED_THROUGH_KEY);
+        }
+
+        return $dictionary;
+    }
+}
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: cs

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+    protected function setSoftDeletingConstraints()
+    {
+        $this->getQuery()->whereNull($this->getParent()->getQualifiedDeletedAtColumn());
+        // Note: Don't know if there is need to check intermediate models.
+    }
+
+    /**
+     * @param \Illuminate\Database\Eloquent\Collection $results
+     *
+     * @return \Illuminate\Database\Eloquent\Model[]
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: just `array` please

Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: laravel doesn't use that phpdoc style

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+     * @return void
+     */
+    protected function setSoftDeletingConstraints()
+    {
+        $this->getQuery()->whereNull($this->getParent()->getQualifiedDeletedAtColumn());
+        // Note: Don't know if there is need to check intermediate models.
+    }
+
+    /**
+     * @param \Illuminate\Database\Eloquent\Collection $results
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: cs

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+     *
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: cs

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+    /**
+     * @return void
+     */
+    protected function setSoftDeletingConstraints()
+    {
+        $this->getQuery()->whereNull($this->getParent()->getQualifiedDeletedAtColumn());
+        // Note: Don't know if there is need to check intermediate models.
+    }
+
+    /**
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: missing short desc

**Comment by znck** [comment]
@GrahamCampbell

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+     * @var string
+     */
+    protected $localKey;
+
+    /**
+     * Create a new instance of relation.
+     *
+     * @param \Illuminate\Database\Eloquent\Builder  $query
+     * @param \Illuminate\Database\Eloquent\Model  $parent
+     * @param \Illuminate\Database\Eloquent\Model[]  $models
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: `array` please

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+     * @param string|null  $localKey
+     */
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: missing `@return void`

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+        ]);
+        $this->getQuery()->whereIn($this->getParent()->getQualifiedKeyName(), $this->getKeys($models, $this->localKey));
+    }
+
+    /**
+     * Initialize the relation on a set of models.
+     *
+     * @param \Illuminate\Database\Eloquent\Model[]  $models
+     * @param string  $relation
+     *
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: no extra line here please

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 20]
```diff
@@ -0,0 +1,184 @@
...
+        return $models;
+    }
+
+    /**
+     * Match the eagerly loaded results to their parents.
+     *
+     * @param array  $models
+     * @param \Illuminate\Database\Eloquent\Collection  $results
+     * @param string  $relation
+     *
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: same

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 21]
```diff
@@ -0,0 +1,184 @@
...
+            'Illuminate\Database\Eloquent\SoftDeletes',
+            class_uses_recursive(get_class($this->getParent()))
+        );
+    }
+
+    /**
+     * Build model dictionary keyed by the relation's foreign key.
+     *
+     * @param \Illuminate\Database\Eloquent\Collection  $results
+     *
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: same

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 22]
```diff
@@ -0,0 +1,184 @@
...
+        $this->getQuery()->addSelect([
+            $this->getParent()->getQualifiedKeyName().' as '.self::RELATED_THROUGH_KEY,
+        ]);
+        $this->getQuery()->whereIn($this->getParent()->getQualifiedKeyName(), $this->getKeys($models, $this->localKey));
+    }
+
+    /**
+     * Initialize the relation on a set of models.
+     *
+     * @param \Illuminate\Database\Eloquent\Model[]  $models
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: `@param  array  $models`

Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: With 2 spaces between each bit

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 23]
```diff
@@ -0,0 +1,184 @@
...
+     * The local key on the relationship.
+     *
+     * @var string
+     */
+    protected $localKey;
+
+    /**
+     * Create a new instance of relation.
+     *
+     * @param \Illuminate\Database\Eloquent\Builder  $query
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: missing space before each

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 24]
```diff
@@ -0,0 +1,184 @@
...
+    /**
+     * Set the required joins on the relation query.
+     *
+     * @return void
+     */
+    protected function setJoins()
+    {
+        $one = $this->getRelated()->getQualifiedKeyName();
+        $prev = $this->getRelated()->getForeignKey();
+        foreach ($this->models as $key => $model) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: don't need to have the `$key =>` bit

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 25]
```diff
@@ -0,0 +1,184 @@
...
+    {
+        $dictionary = $this->buildDictionary($results);
+
+        foreach ($models as $model) {
+            $key = $model->{$this->localKey};
+
+            if (isset($dictionary[$key])) {
+                $val = $dictionary[$key];
+
+                $model->setRelation($relation, $val);
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: please just write `$model->setRelation($relation, $dictionary[$key])`

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 26]
```diff
@@ -0,0 +1,184 @@
...
+            $model->setRelation($relation, $this->getRelated());
+        }
+
+        return $models;
+    }
+
+    /**
+     * Match the eagerly loaded results to their parents.
+     *
+     * @param array  $models
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: 2 spaces needed after `@param`, not 1 please

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 27]
```diff
@@ -0,0 +1,184 @@
...
+
+        $this->getQuery()->select([$this->getRelated()->getTable().'.*']);
+
+        if (static::$constraints) {
+            $this->getQuery()->where($this->getQualifiedParentKeyName(), '=', $this->parent[$this->localKey]);
+        }
+
+        if ($this->parentSoftDeletes()) {
+            $this->getQuery()->whereNull($this->getParent()->getQualifiedDeletedAtColumn());
+            // Note: Don't know if there is need to check intermediate models.
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: this needs addressing and then the comment removing

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 28]
```diff
@@ -0,0 +1,184 @@
...
+
+use Illuminate\Database\Eloquent\Builder;
+use Illuminate\Database\Eloquent\Collection;
+use Illuminate\Database\Eloquent\Model;
+
+class BelongsToThrough extends Relation
+{
+    /**
+     * Column alias for matching eagerly loaded models.
+     */
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: please include `@var string` in the phpdoc

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 29]
```diff
@@ -0,0 +1,184 @@
...
+    const RELATED_THROUGH_KEY = '__deep_related_through_key';
+
+    /**
+     * List of intermediate model instances.
+     *
+     * @var \Illuminate\Database\Eloquent\Model[]
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: please just `array` here

Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: laravel doesn't use the `[]` syntax

File: src/Illuminate/Database/Eloquent/Model.php [hunk 30]
```diff
@@ -838,6 +839,38 @@ public function belongsTo($related, $foreignKey = null, $otherKey = null, $relat
     }
 
     /**
+     * Define a belongs-to-through relationship.
+     *
+     * @param string  $related
+     * @param string|array  $through
+     * @param string|null  $localKey
+     *
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: no extra line please

**Comment by GrahamCampbell** [comment]
I've added a load more comments. Please squash to one commit when done. :)

**Comment by znck** [comment]
@GrahamCampbell  Is there a way to automatically fix doc comments? 
and phpcs --standard=PSR2 passes but StyleCI fails. any fix?

**Comment by GrahamCampbell** [comment]
We don't use phpcs. Just commit your code, then download and apply the diff from StyleCI. Alternatively, you can run php-cs-fixer with the config committed to this repo, and you'll get nearly identical results to StyleCI.

In terms of the phpdoc, the automated fixers only partially reflect laravel's code style. I'm yet to write a full fixer suite.

File: src/Illuminate/Database/Eloquent/Model.php [hunk 31]
```diff
@@ -838,6 +839,37 @@ public function belongsTo($related, $foreignKey = null, $otherKey = null, $relat
...
+     * @return \Illuminate\Database\Eloquent\Relations\BelongsToThrough
+     */
+    public function belongsToThrough($related, $through, $localKey = null)
+    {
+        $related = new $related;
+        $models = [];
+        foreach ((array) $through as $key => $model) {
+            $object = new $model;
+            if (! ($object instanceof self)) {
+                throw new \InvalidArgumentException(
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: you MUST import this

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 32]
```diff
@@ -0,0 +1,191 @@
...
+     *
+     * @param  array  $models
+     * @return void
+     */
+    public function addEagerConstraints(array $models)
+    {
+        $this->getQuery()->addSelect([
+            $this->getParent()->getQualifiedKeyName().' as '.self::RELATED_THROUGH_KEY,
+        ]);
+        $this->getQuery()->whereIn($this->getParent()->getQualifiedKeyName(), $this->getKeys($models, $this->localKey));
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: missing newline before this line

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 33]
```diff
@@ -0,0 +1,191 @@
...
+    }
+
+    /**
+     * Initialize the relation on a set of models.
+     *
+     * @param  \Illuminate\Database\Eloquent\Model[]  $models
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: array please

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 34]
```diff
@@ -0,0 +1,191 @@
...
+    /**
+     * Determine whether the model uses Soft Deletes.
+     *
+     * @param  \Illuminate\Database\Eloquent\Model  $model
+     * @return bool
+     */
+    public function hasSoftDeletes(Model $model)
+    {
+        return in_array(
+            'Illuminate\Database\Eloquent\SoftDeletes',
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: maybe better to import this, then write `return in_array(SoftDeletes::class, class_uses_recursive(get_class($model)));`

File: src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php [hunk 35]
```diff
@@ -0,0 +1,191 @@
...
+     * @param  \Illuminate\Database\Eloquent\Collection  $results
+     * @return array
+     */
+    protected function buildDictionary(Collection $results)
+    {
+        $dictionary = [];
+
+        foreach ($results as $result) {
+            $dictionary[$result->{static::RELATED_THROUGH_KEY}] = $result;
+            $result->offsetUnset(static::RELATED_THROUGH_KEY);
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php]: does calling unset not work?

**Comment by taylorotwell** [comment]
Can you post the table structure for your original problem in the original post of that issue thread?

**Comment by taylorotwell** [comment]
Closing since 2 weeks of inactivity. If someone wants to open it back up go for it.

**Comment by blomqvist** [comment]
I can post an example:

**The model table**
```
order_items:
 - id
 - product_id
 - data_2
 - data_3
 - data_4
```

**The pivot table**
```
order_item_order_item:
 - contractor_order_item_id (unique)
 - customer_order_item_id
 - customer_order_item_special_attribute
```

**The relation is like this:**
A contractor order item only has one customer order item
A customer order item can have many contractor order items
The data stored in the pivot table is needed to be able to pinpoint which special attributes belongs to which customer order item.

It is not possible to have the `customer_order_item_special_attribute` on the order_item table as the attribute value might be different depending on which `contractor_order_item` is connected to it.

Does this make sense? :smile:

**Comment by EliasZ** [comment]
ping @znck

**Comment by znck** [comment]
@EliasZ  Use this https://github.com/znck/belongs-to-through

**Comment by canfiax** [comment]
+1 this is really needed.

Also described here: https://laracasts.com/discuss/channels/eloquent/has-model-through-relation

Say I have the following relationships:

       A Company has many Plan
       A Plan belongs to one Company
       A User has many Subscription
       A Subscription belongs to one Plan

How can I get the `Company` from a `Subscription`? I could do something like this:

      $subscription->plan->company

But I would like to do:

     $subscription->company

and:

    Subscription::with('company')->get();

**Comment by Evertt** [comment]
@znck why have you never completed this PR? It'd be very nice to have this integrated into eloquent itself.

**Comment by znck** [comment]
I can resend a PR (Next week) if it something people really need.

**Comment by znck** [comment]
@taylorotwell What do you think? Should it be in Eloquent or independent package is okay?

**Comment by fernandobandeira** [comment]
There's an open PR for this already, check #17417 😄

**Pull request closed without merging**